shakytweaks/src/main/java/com/mindsea/shakytweaks/TweakValueResolver.kt

```diff
@@ -111,4 +122,126 @@ internal class TweakValueResolver(
             else -> throw UnsupportedOperationException("Tweak $tweak doesn't support increment")
         }
     }
+
+    fun allowsDecimals(key: String): Boolean {
+        return when (val tweak = tweakProvider.findTweak(key)) {
+            is Tweak.IntTweak -> false
+            is Tweak.LongTweak -> false
+            is Tweak.FloatTweak -> true
+            is Tweak.DoubleTweak -> true
+            else -> throw UnsupportedOperationException("Tweak $tweak isn't a number")
+        }
+    }
+
+    fun allowsNegativeNumbers(key: String): Boolean {
+        return when (val tweak = tweakProvider.findTweak(key)) {
+            is Tweak.IntTweak -> tweak.minValue < 0
+            is Tweak.LongTweak -> tweak.minValue < 0
+            is Tweak.FloatTweak -> tweak.minValue < 0
+            is Tweak.DoubleTweak -> tweak.minValue < 0
+            else -> throw UnsupportedOperationException("Tweak $tweak isn't a number")
+        }
+    }
+
+    fun updateValueFromString(key: String, string: String) {
+        when (val tweak = tweakProvider.findTweak(key)) {
+            is Tweak.IntTweak -> {
+                val updatedValue = string.toInt()
+                when {
+                    updatedValue < tweak.minValue -> {
+                        throw TweakValueOutOfRangeException(
+                            "The value $updatedValue is smaller than the minimum value (${tweak.minValue})",
+                            TweakValueOutOfRangeException.ThresholdType.MIN,
+                            string,
+                            tweak.minValue.toString()
+                        )
+                    }
+                    updatedValue > tweak.maxValue -> {
+                        throw TweakValueOutOfRangeException(
+                            "The value $updatedValue is larger than the maximum value (${tweak.maxValue})",
+                            TweakValueOutOfRangeException.ThresholdType.MAX,
+                            string,
+                            tweak.maxValue.toString()
+                        )
+                    }
+                    else -> {
+                        updateValue(key, updatedValue)
```

Comment:
Try using a function with generics to reduce these duplicate codes. Use `T` as the type you will have to map to.

It looks bad that tweaks them selves don't have generics nor common interfaces, but still it is possible to make a generic helper function

```
   fun <T> stringNumberValueChecker(
        key: String,
        string: String,
        minValue: T,
        maxValue: T,
        mapper: (String) -> T
    ) {

        val updatedValue = mapper(string)
        when {
            updatedValue < minValue -> {
                throw TweakValueOutOfRangeException(
                    "The value $updatedValue is smaller than the minimum value (${tweak.minValue})",
                    TweakValueOutOfRangeException.ThresholdType.MIN,
                    string,
                    tweak.minValue.toString()
                )
            }
            updatedValue > maxValue -> {
                throw TweakValueOutOfRangeException(
                    "The value $updatedValue is larger than the maximum value (${tweak.maxValue})",
                    TweakValueOutOfRangeException.ThresholdType.MAX,
                    string,
                    tweak.maxValue.toString()
                )
            }
            else -> {
                updateValue(key, updatedValue)
            }

        }
    }
```

Then, call it like this:
```
   fun updateValueFromString(key: String, string: String) {
        when (val tweak = tweakProvider.findTweak(key)) {
            is Tweak.IntTweak -> {
                stringNumberValueChecker(key, string, tweak.minValue, tweak.maxValue) { it.toInt() }
            }
            // ...
        }
    }
```